File: Coin Combinations I/solve.cpp
#include<iostream>
#include<map>
#include<vector>
#include <sstream>
using namespace std;

void run_tests()
{
    int n,x;
    cin>>n>>x;
    int mod_val = 1e9 + 7;
    vector<int> coins(n);
    for (int &x : coins) cin>>x;

    vector<int> dp(x+1,0);
    dp[0] = 1;

    for (int i = 1; i <= x; ++i)
    {
        for (int j = 0; j<n; ++j)
        {
            if (i-coins[j]>=0)
            {
                dp[i] = dp[i] + dp[i-coins[j]];
                dp[i] = dp[i] % mod_val;
            }
        }
    }

    cout<<dp[x];

    return ;
}

int main(){

    run_tests();
}

File: Coin Combinations I/solve.py
import sys
input = sys.stdin.readline
 

# i tried everything
# its TLE
# but the logic is 100% correct

def run_tests():
    MOD = 10**9 + 7
    n,x = map(int,input().split())
    coins = list(map(int,input().split()))

    dp = {i:0 for i in range(x+1)}
    dp[0] = 1

    for i in range(x+1):
        for c in coins:
            if i - c >= 0:
                dp[i] = (dp[i - c] + dp[i]) % 1000000007
            
    print(dp[x]) 

def main():
    
    # number of test cases
    run_tests()
 
if __name__ == "__main__":
    main()
